Bringing the latest updates

atcnthn16 opened the pull request:
Bringing the updates from the local Strata (.zip) of Rebel.
Use MKL if installed in the system, fall back to OpenBLAS otherwise.

atcnthn16 committed:
updated source code and modified CMakeLists.txt

atcnthn16 committed:
use MKL if installed, OpenBLAS otherwise

atcnthn16 commented:
will reopen after bringing the use of yaml-cpp from the system

atcnthn16 committed:
link to the installed yaml-cpp

shash-sharma commented on inc/MGF.hpp:
```diff
@@ -117,6 +122,17 @@ class MGF
 	// ============ Interface ============
 
     void Initialize(double _f, LayerManager &_lm, MGF_settings &_s);
+    void adaptiveInterpolation(MGF &mgf);
```
Not a big deal, but in the rest of the code `CamelCase` was used while here `lowerCamelCase` is used. Might be good to stay consistent? Similar for the use of `_` in some method names.

shash-sharma commented on src/MGF.cpp:
```diff
@@ -1013,6 +1418,198 @@ void MGF::TabulateMGF(std::vector<std::vector<table_entry<N>>> &table, bool curl
+
+}
+
+template<std::size_t N>
```
Maybe I missed it, but to have template functions defined in the cpp file (rather than the hpp file) usually one has to specialize the template in the cpp file. I don't see where these templates have been specialized. How did you manage to get the code to compile?

shash-sharma commented on inc/MGF.hpp:
```diff
@@ -148,6 +164,9 @@ class MGF
 	template<std::size_t N>
 	void ComputeMGF_Interpolation(double rho, double z, double zp, std::array<std::complex<double>, N> &G, std::vector<std::vector<table_entry<N>>> &table, std::vector<bool> &components);
 
+    template<std::size_t N>
```
please see the question about template specialization in the cpp file

shash-sharma commented on inc/layers.hpp:
```diff
@@ -48,17 +49,19 @@ class Layer
 	int layerID;
 	double zmax, zmin, h, epsr, sigma, mur, sigmamu;
 	std::vector<int> objectIDs;   /// List of indices of objects that belong to this layer
+	double epsr_im; //Imaginary part of epsilon, indicating a frequency independent loss
```
probably can be frequency-dependent just as well?

shash-sharma replied:
just curious, why not make epsr `std::complex<double>` instead?

shash-sharma commented on src/DCIM.cpp:
```diff
@@ -51,7 +51,11 @@
 #include <stdexcept>
 
-#include <cblas.h>
+#if defined(USE_MKL)
```
curious why mkl support was needed? Do you see a performance difference, or is it just to avoid having to install openblas if one already has mkl?

shash-sharma commented on src/spectral_MGF.cpp:
```diff
@@ -864,6 +864,17 @@ void SpectralMGF::ComputeKmi(std::array<std::complex<double>, 5> &K)
 		
 		K[2] = ((km*km/(kzm*kzm))*WIe_mi - WIh_mi)/krhosq;
+
+		/*
```
For this and the commented block below, maybe add a note at the top like "for debugging"

shash-sharma commented on test/CMakeLists.txt:
```diff
@@ -17,8 +17,7 @@
 # You should have received a copy of the GNU General Public License
 # along with Strata.  If not, see <https://www.gnu.org/licenses/>.
 
-set(STRATA_LIB
```
I don't remember why I had it this way previously, but might be worth testing on a couple platforms just to be sure it works? E.g. your local computer and on a workstation you may have access to?

shash-sharma commented on CMakeLists.txt:
```diff
@@ -57,30 +57,42 @@ set(WITH_OPENBLAS_LIB "" CACHE PATH "Path to an existing OpenBLAS static library
-    set(OPENBLAS_INC "${OPENBLAS_ROOT}/include")
+# ====== Find MKL ======
+
+set(MKL_INTERFACE lp64)
```
I'm not familiar with CMake-mkl interplay, but might be worth checking that these options don't lead to issues on other platforms? Or in the case that mkl is found, but openmpi is not found? Etc.

shash-sharma commented on src/layers.cpp:
```diff
@@ -461,24 +461,99 @@ void LayerManager::ProcessLayers(double f)
+        std::vector<double> z_nodes_local;
+
+        double h = layers[ii].zmax - layers[ii].zmin;
+        double lambda = lambda0/std::sqrt(layers[ii].epsr);
```
Might be better to calculate `lambda` taking into account the imaginary part of epsr and the conductivity - both would cause the wavelength to be smaller in the layer than for the purely lossless case. You could instead compute the complex wave number and use its real part to calculate wavelength

shash-sharma commented on src/layers.cpp:
```diff
@@ -461,24 +461,99 @@ void LayerManager::ProcessLayers(double f)
+    int idx_layer = FindLayer(z_min);
+
+    double h = z_max - z_min;
+    double lambda = lambda0/std::sqrt(layers[idx_layer].epsr);
```
same thing as above for `lambda` here

shash-sharma commented on inc/MGF.hpp:
```diff
@@ -117,6 +122,17 @@ class MGF
+    void adaptiveInterpolation(MGF &mgf);
+    void updateZnodes(MGF &mgf);
+    void updateRhonodes(MGF &mgf);
+    bool isMidpointCorrect(double rho, double z_src, double z_test,  double adaptive_threshold, bool test_rho);
```
Would be worth thinking about which of these many new methods needs to be accessed by an external user, and should thus be considered a part of the API? Those that are internal helper methods, i.e. never meant to be called by a user of Strata but just used internally by other methods, should be placed in a `protected` or `private` block near the bottom of the class definition, imo, so that users are not overwhelmed when they look at the header

shash-sharma commented on inc/MGF.hpp:
```diff
@@ -117,6 +122,17 @@ class MGF
+    void addZTable(MGF &mgf, double z_test, int layer);
+    void plotRhoNodes(MGF mgf, std::vector<double> z_gridpoints);
+    void processRhoTests(MGF& mgf, double rho_spacing, double z_test, double z_src, int level, std::vector<double>& rho_tests);
+    void processZTests(MGF& mgf, int layer_idx, double rho_test, double z_spacing, double z_test, double z_src, int level, std::vector<double>& z_tests);
```
In all these new methods, why take `MGF& mgf` as an argument rather than just directly modifying `this` object? Or else, if the methods do not depend on `this` object, then they should be marked `static`

shash-sharma reviewed (commented):
For the mgf.cpp file, I didn't follow the logic of the interpolations too closely since there was a lot of code, but seems like a lot of useful functionality was added! I've added my comments. Thanks!

pierotriverio commented:
Thanks @shash-sharma! @atcnthn16 can you have a look and respond?

atcnthn16 committed:
check if the libraries are available on the system, added an option to build the documentation

atcnthn16 committed:
complex epsr, internal methods moved to private, corrections on lambda, CMakeLists.txt modified

Commit:
corrected epsr definition and revised lambda calculation with complex epsilon

atcnthn16 commented:
Dear Prof. @shash-sharma,

Thank you so much for your feedback. I've made the following modifications:

- Used ```CamelCase``` for the new methods, and moved the "Computational drivers" to private as they're only internally called.
- Changed epsr variable to be complex, and modified lambda accordingly to include both the conductivity and the imaginary part of it.

About your other comments:

- CMake/MKL: We are trying to containerize Rebel which uses MKL, so we just want to use the same BLAS library. We use the same CMake convention for Rebel, and it works with it in different environments. For Strata, the code downloads (or links to the one indicated with flags) OpenBLAS if MKL is not found.
- Templates: The code compiles as we instantiate the templated methods in the .cpp file by calling them from a non-templated method:
      - Methods with ```table_entry<N>``` are called with the instantiated (with N = 5 and 4) [```MGF_table``` and ```CurlMGF_table```](https://github.com/modelics/strata/blob/apptainer/inc/MGF.hpp#L181-L182) variables.
      - ```ComputeMGF``` methods with ```std::array<std::complex<double>, N> &G``` are also called with locally instantiated G variables, can be seen in [```isMidpointCorrect```](https://github.com/modelics/strata/blob/apptainer/src/MGF.cpp#L381-L395) example.
- MGF.cpp logic: We initialize the interpolation table with the z gridpoints (minimum 2 points to define the boundaries) and rho nodes, then we perform ```AdaptiveInterpolation``` which updates the interpolation grid by checking the value in the midpoints by comparing the integrated values with the interpolated ones. The code with new functionalities and helper methods were implemented by Yongzhong, who can provide better explanations. (Dear Prof. @pierotriverio, does Yongzhong have access to this conversation?)

Regards,
Atacan

pierotriverio commented:
Thanks @atcnthn16! Adding @yongzhongli  to the discussion

Commit:
enable OpenMP in TabulateMGF function, add a user-defined variable to control the adaptive sampling along z direction

Commit:
add a test script to the new adaptive sampling strategy

Commit:
Reduce verbose logging during adaptive sampling tests

yongzhongli commented:
Thanks @atcnthn16 for the modifications, I can give more explanations on some of the functionalities in this PR @shash-sharma @pierotriverio. This updates include the following improvement proposed for strata:

-  The adaptive sampling of MGF: The initial interpolation table built for MGF may be too coarse to achieve sufficient accuracy or too fine which adds extra computation. We choose to use a small interpolation table with minimum 2 nodes defined on the boundary of each dielectric layer and some rho nodes determined by a minimum number of sampling points per wavelength. Then we use [`AdapativeInterpolation`](https://github.com/modelics/strata/blob/fafd240826a859670a30dd9302fc3efa1e66acdc/src/MGF.cpp#L236) function to update the initial interpolation table by testing the interpolation at new nodes in the $\rho$ and $z$ direction. The rules to add new nodes are as follows,
    - The new nodes are the midpoints of two existing nodes
    - We compare the interpolated MGF and integrated ones at these midpoints. If their relative difference is below the error threshold, we move to the next one until all the midpoints are traversed. If not, we will add that midpoint to the interpolation table and continue the test.
- Fixed the indexing issues when there are two coincident nodes at the layer interface
- OpenMP parallelization: The [`TabulateMGF`](https://github.com/modelics/strata/blob/fafd240826a859670a30dd9302fc3efa1e66acdc/src/MGF.cpp#L1308) function is paralleled using OpenMP directives to improve the multi-threading performance 

I made some new modifications to the working branch to enable OpenMP, add a test script for the new adaptive sampling strategy, and remove some verbose logging.

pierotriverio commented:
Thanks @yongzhongli ! @shash-sharma can you have a look when you get a chance?

Commit:
Update the parallelization strategy for TabulateMGF function

Commit:
Set smaller threshold for adaptive interpolation

shash-sharma reviewed (approved):
Sorry about the delay, it seems I stopped receiving email notifications from GitHub, which I've fixed. Thanks for the explanations about interpolation and compilation, looks great to me.
